src/mscorlib/shared/System/IO/PathHelper.Windows.cs

```diff
@@ -24,63 +25,73 @@ internal class PathHelper
         /// <returns>Normalized path</returns>
         internal static string Normalize(string path)
         {
-            // Get the full path
-            StringBuffer fullPath = new StringBuffer(path.Length);
+            if (path.Length < PathInternal.MaxShortPath)
+            {
+                Span<char> initialBuffer = stackalloc char[PathInternal.MaxShortPath];
+                return Normalize(path, new ValueStringBuilder(initialBuffer));
+            }
+            else
+            {
+                return Normalize(path, new ValueStringBuilder());
+            }
+        }
 
+        private static string Normalize(string path, ValueStringBuilder builder)
+        {
+            // Get the full path
             try
             {
-                GetFullPathName(path, ref fullPath);
+                GetFullPathName(path, ref builder);
 
-                if (fullPath.Contains('~'))
+                if (builder.AsSpan().Contains('~'))
                 {
-                    return TryExpandShortFileName(ref fullPath, originalPath: path);
+                    return TryExpandShortFileName(ref builder, originalPath: path);
                 }
                 else
                 {
-                    if (fullPath.Length == path.Length && fullPath.StartsWith(path))
+                    if (builder.AsSpan().Equals(path.AsReadOnlySpan()))
                     {
                         // If we have the exact same string we were passed in, don't bother to allocate another string from the StringBuffer.
                         return path;
                     }
-                    return fullPath.ToString();
+                    return builder.ToString();
                 }
             }
             finally
             {
                 // Clear the buffer
-                fullPath.Free();
+                builder.Dispose();
             }
         }
 
-        private static unsafe void GetFullPathName(string path, ref StringBuffer fullPath)
+        private static void GetFullPathName(string path, ref ValueStringBuilder builder)
         {
             // If the string starts with an extended prefix we would need to remove it from the path before we call GetFullPathName as
             // it doesn't root extended paths correctly. We don't currently resolve extended paths, so we'll just assert here.
             Debug.Assert(PathInternal.IsPartiallyQualified(path) || !PathInternal.IsExtended(path));
+            builder.EnsureCapacity(path.Length);
+            builder.Append('\0');
```

Comment:
Why are we appending the zero? It does not seem necessary.

Reply:
I could set the length to 1. We have an assert that doesn't allow you index beyond the length. Thoughts?

Reply:
You can use `Marshal.GetReference(...)` to get the pointer instead of taking address of the element zero. E.g. Add `ref char GetPinnableReference() => Marshal.GetReference(_chars)` method to the ValueStringBuilder. Use that instead of taking address of element zero. It will be also a bit smaller code because of it will avoid the useless boundscheck against zero.

Reply:
Will do that, thanks.